python/src/iceberg/table/schema.py

```diff
@@ -0,0 +1,286 @@
+        ...
+
+    @abstractmethod
+    def field(self, field: NestedField, field_result: T) -> T:
+        ...
+
+    @abstractmethod
+    def list(self, list_type: ListType, element_result: T) -> T:
+        ...
+
+    @abstractmethod
+    def map(self, map_type: MapType, key_result: T, value_result: T) -> T:
+        ...
+
+    @abstractmethod
+    def primitive(self, primitive: PrimitiveType) -> T:
+        ...
+
+
+def visit(obj, visitor: SchemaVisitor[T]) -> T:
+    if isinstance(obj, Schema):
+        return visitor.schema(obj, visit(obj.as_struct(), visitor))
+
+    elif isinstance(obj, StructType):
+        results = []
+        for field in obj.fields:
+            visitor.before_field(field)
+            try:
+                result = visit(field.type, visitor)
+            finally:
+                visitor.after_field(field)
+
+            results.append(visitor.field(field, result))
+
+        return visitor.struct(obj, results)
+
+    elif isinstance(obj, ListType):
+        visitor.before_list_element(obj.element)
+        try:
+            result = visit(obj.element.type, visitor)
+        finally:
+            visitor.after_list_element(obj.element)
+
+        return visitor.list(obj, result)
+
+    elif isinstance(obj, MapType):
+        visitor.before_map_key(obj.key)
+        try:
+            key_result = visit(obj.key.type, visitor)
+        finally:
+            visitor.after_map_key(obj.key)
+
+        visitor.before_map_value(obj.value)
+        try:
+            value_result = visit(obj.value.type, visitor)
+        finally:
+            visitor.after_list_element(obj.value)
+
+        return visitor.map(obj, key_result, value_result)
+
+    elif isinstance(obj, PrimitiveType):
+        return visitor.primitive(obj)
+
+    else:
+        raise NotImplementedError("Cannot visit non-type: %s" % obj)
+
+
+def index_by_id(schema_or_type) -> Dict[int, NestedField]:
+    class IndexById(SchemaVisitor[Dict[int, NestedField]]):
+        def __init__(self) -> None:
+            self._index: Dict[int, NestedField] = {}
+
+        def schema(self, schema, result):
+            return self._index
+
+        def struct(self, struct, results):
+            return self._index
+
+        def field(self, field, result):
+            self._index[field.field_id] = field
+            return self._index
+
+        def list(self, list_type, result):
+            self._index[list_type.element.field_id] = list_type.element
+            return self._index
+
+        def map(self, map_type, key_result, value_result):
+            self._index[map_type.key.field_id] = map_type.key
+            self._index[map_type.value.field_id] = map_type.value
+            return self._index
+
+        def primitive(self, primitive):
+            return self._index
+
+    return visit(schema_or_type, IndexById())
+
+
+def index_by_name(schema_or_type) -> Dict[str, int]:
+    class IndexByName(SchemaVisitor[Dict[str, int]]):
```

Comment:
It looks like this doesn't implement short names like the Java class does. Is it possible to add short names? It's nice to be able to refer to `locations.latitude` rather than `locations.element.latitude`.

Reply:
Added the logic for short-names. Thanks for the help with this!